### DISCUSSION TITLE
Flux Monitor normal vector in 2D simulation

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2442] [Q&A] [answered]
Any meep devs out there:

Let's say I'm running a 2D simulation, would there be any cases where it would make sense for the Poynting vector to have a nonzero component in the z-direction? Or rather, in a case where there is a Poynting vector which has a nonzero component in the z-direction, does it make sense to have a 2D simulation?

Here is a very simple example with a wonderful paint drawing I did on my touchpad:

![image](https://user-images.githubusercontent.com/37753156/225132872-a1fdcfe0-829d-4220-a451-9a2faa9dd9bd.png)

You have a plane wave propagating along the x-z direction with the Poynting vector in the (1,0,1) direction, the geometry is translationally symmetric along the z-axis. Would there be any case where one would use a 2D simulation?

(I might resurrect the poynting flux adjoint monitor and I want to know whether this is a case to include in the code or not.)

**Comment by Andeloth** [comment]
I just realized that Meep currently doesn't have arbitrary directional capabilities for flux monitors, so this question is irrelevant at this time.

**Comment by smartalecH** [comment]
Can you describe exactly what you want to optimize? This is a bit vague...

**Reply by Andeloth** [reply]
Not optimize, I was thinking about whether it makes sense to ever have a flux monitor in a non-planar direction in a 2D meep optimization. I'm working on the PoyntingFlux adjoint implementation and I was thinking about arbitrary directionality for the monitors, e.g. having a flux monitor normal to the (1,2,3) vector in 3D. I thought that maybe there could be a case in a 2D simulation where we'd want to know the flux at an angle with respect to the simulation plane. Then I realized that meep doesn't actually support monitors along non-Cartesian axis directions which makes this question irrelevant.

Though I suppose you could technically implement a flux monitor along the z direction in a 2D simulation? Though I'm not sure whether that makes much sense to do. What do you think, Alec?

**Reply by Andeloth** [reply]
Alternatively, does it make any sense to have a 1D flux region (a line) in a 3D simulation?

**Reply by smartalecH** [reply]
> I'm working on the PoyntingFlux adjoint implementation

I would coordinate with @Alex-Kaylor. He got pretty far with his implementation (just debugging some issues with 3D IIRC). No need to reinvent the wheel here (especially because there are so many nuances with the Yee grid etc).

> Though I suppose you could technically implement a flux monitor along the z direction in a 2D simulation?

Sure you could do this. I could see this being useful when you have [a non-zero `k` in the z-direction](https://meep.readthedocs.io/en/latest/2d_Cell_Special_kz/) (although this requires you flip the sign of `k` here during the adjoint run... which we haven't set up yet...)

> Alternatively, does it make any sense to have a 1D flux region (a line) in a 3D simulation?

There's not a normal direction associated with this case, so I would ignore it. (I'm sure you could make it work if you really wanted to stretch the math and interpretation of things...)

**Reply by Andeloth** [reply]
I saw @Alex-Kaylor 's PR (#2191) and I'm working with that as the baseline. I thought that the consensus was that we wanted a C++ implementation of the flux calculation to ensure the fields from the entire cell aren't loaded into all chunks at the end of the simulations, so I was changing the implementation based on that.

**Comment by stevengj** [comment, marked as answer]
> Not optimize, I was thinking about whether it makes sense to ever have a flux monitor in a non-planar direction in a 2D meep optimization.

It depends on what you mean by "2D". 

If by "2D" you mean "z-invariant geometry" *and* z-invariant fields (i.e. kz = 0), then by construction the Poynting flux in the z direction will always be zero (at least with reciprocal materials … for non-reciprocal anisotropic materials I should be careful about glib statements because things can get hairy).

If by "2D" you mean "z-invariant geometry" but there are z-varying fields (i.e. kz ≠ 0) — see the tutorial [2d cell with out-of-plane wavevector](https://meep.readthedocs.io/en/latest/2d_Cell_Special_kz/) — then in that case there can be Poynting flux in the z direction, and it might be useful to calculate in some cases (e.g. to get group velocity in z of a time-harmonic mode, since in that case group velocity is z Poynting flux divided by energy density).
